### DISCUSSION TITLE
Access current state in action / aspect programing on state entry - exit

### DISCUSSION THREAD

**Discussion by ronanquillevere** [statelyai/xstate#1294] [General] [answered]
Hello,  first thanks for the awesome lib.

I have not found a way to access the current state easily  inside an action. Of course I could store it inside the context then retrieve it from the context inside my action but this feels really bad. Am I missing something ? Especially for entry and exit actions that are not triggered by events, that would be useful imo. 

**What I am trying to do**
I am  simply trying to record how much time I spend in each state. I use the entry and exit actions to get times and then I compute the time spent for each state. I would like to always call the **same method** on entry/exit for each state and associate the times with the state name one way or another. 

```js
myState: {
      entry: ['setStartTime'],
      exit: ['setEndTime'],
...
},
myState2: {
      entry: ['setStartTime'],
      exit: ['setEndTime'],
...
},
```
Yes I could create distincts methods for each entry/exit/state that would call the same method underneath to avoid duplicating the code, but I was wondering if there was a way to do this kind of aspect programming stuff.

If someone has a better idea let me know :)

**Comment by pipex** [comment, marked as answer]
I'm learning about xstate as well, but maybe you could use the onTransition event on an [interpreter](https://xstate.js.org/docs/guides/interpretation.html#interpreting-machines), something like

```javascript
let currentState;
const service = interpret(machine).onTransition(state => {
   if (state.value != currentState) {
       // TODO: terminate timer if any and start a new one
      currentState = state.value;
   }
});
```

**Reply by davidkpiano** [reply]
Yep, `.onTransition(...)` would be the way to go. See this example: https://codesandbox.io/s/timing-transitions-example-j1ojp?file=/src/index.js

```js
let lastTransitionTime;
const service = interpret(machine).onTransition(state => {
  if (state.changed === undefined) {
    lastTransitionTime = Date.now();
  }

  if (state.history && !state.history.matches(state.value)) {
    const now = Date.now();
    console.log(
      `Time spent in ${state.history.value}: ${now - lastTransitionTime}ms`
    );
    lastTransitionTime = now;
  }
});
```

**Reply by vck3000** [reply]
I have a similar use case except I need to be able to modify context and state (by firing events). Any suggestions on how could I go about doing this?

Essentially I want a function to be called on each transition. If a certain set of criteria are met, I want it to modify the context and state. I have used a bit of `xstate.actions.pure()` and that has worked great but I can't access the current state. 

I also tried to use an assign action on entry to specific states, but the assign function's `assignMeta` data holds the previous (non-transitioned) state's information instead of the state that the machine has just entered. 

I could also just store the state value in context on every transition and access it that way (as the author pointed out too). 

Alternatively, should I go external and use `onTransition()`? 

Thanks :)

Edit: Sorry, it appears assignMeta holding previous state was already discussed [here](https://github.com/davidkpiano/xstate/discussions/1309). I'll make some extra actions like the author of that post did.

**Reply by Andarist** [reply]
Actually i think the biggest issue with what you are trying to do is not that you get the previous state in the meta argument but rather that it's not easy to express „fire after each transition” logic within the machine.

The closest you can get to it is defining an `always` transition (a transient one) on the config's root. It would be selected after each regular transition not followed by inner transient transitions or after all consecutive inner transient transitions. This should be OK in most cases though.

The problem is that transient transitions are always matched against, even after other transient transitions. So it would have to be guarded and preferrably matching it should invalidate its condition to avoid infinite loops. 

It would also probably resolve your „previous state” problem as in a transient transition the „previous state” should be the one you are looking for - the one to which machine got after preceeding (in most cases regular) transition

**Reply by vck3000** [reply]
Interesting! Eventless (transient) transitions that are guarded to only run 'once' per state change to prevent infinite loops. 

Storing the state as a string in context has worked well for me at the moment. I also (temporarily) solved my issue of "fire after a transition" problem by doing an external transition and adding an `entry` action on the required state node. But if I start to get more states that require similar checks, I'll give the root `always` transitions a shot :). 

Thanks for your help!

**Comment by egor-a-trubnikov-panov** [comment]
I use x-state and x-state vizualizator as a convenient level logic editor in the game. And getting the state name in event calls has become a real problem. Because events are bound via `.withConfig` to different machines. 
And there can be several parallel states in machines.I need a more reliable way to get the state that triggered the event.
The name of the state in my case carries a strong semantic load that determines the logic in the event

**Comment by gbcreation** [comment]
Hello,

In the [documentation about Actions](https://xstate.js.org/docs/fr/guides/actions.html#declarative-actions), it's indicated that the `exec` action implementation function takes three arguments `context`, `event` and `actionMeta`. And the `actionMeta` object contains the `action` and `state` properties.  The doc specifies that this `state` property is the resolved machine state, after transition. From my understanding, the new state could be retrieved from this property.

However, from my tests, `actionMeta` only contains the `action` and `_event` properties. There is no `state` property. Am i doing something wrong or is the doc outdated?

**Reply by davidkpiano** [reply]
The doc may be outdated. We are working on new XState v5 beta docs now.

**Reply by zara-khaliq** [reply]
The docs state that if predictableActionArguments is set to true then you cannot access the state in actionMeta
